### PR TITLE
persist signal accounts

### PR DESCRIPTION
**Pull request by michaelkirk**
Fixes seeing an empty contact list when launching compose view too soon.
Fixes the flickering contact avatar seen when first launching Signal.

PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/ViewControllers/OWSConversationSettingsViewController.m [hunk 1]
```diff
@@ -171,7 +171,7 @@ - (BOOL)hasExistingContact
     OWSAssert([self.thread isKindOfClass:[TSContactThread class]]);
     TSContactThread *contactThread = (TSContactThread *)self.thread;
     NSString *recipientId = contactThread.contactIdentifier;
-    return [self.contactsManager.lastKnownContactRecipientIds containsObject:recipientId];
+    return [self.contactsManager signalAccountForRecipientId:recipientId] != nil;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OWSConversationSettingsViewController.m]: We do this a few places.  I'd make a `hasSignalAccountForRecipientId` method personally to yield more readable code, but I know many people prefer to avoid the redundancy. TIOLI

File: Signal/src/contact/OWSContactsManager.m [hunk 2]
```diff
@@ -240,170 +232,76 @@ - (void)updateSignalAccounts
...
-            }];
+        // Update cached SignalAccounts on disk
+        [self.dbWriteConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+            NSArray<NSString *> *allKeys = [transaction allKeysInCollection:[SignalAccount collection]];
+            NSMutableSet<NSString *> *orphanedKeys = [NSMutableSet setWithArray:allKeys];
+            
+            for (SignalAccount *signalAccount in signalAccounts) {
+                // TODO only save the ones that changed
+                [orphanedKeys removeObject:signalAccount.uniqueId];
+                [signalAccount saveWithTransaction:transaction];
```
Comment (charlesmchen) [inline Signal/src/contact/OWSContactsManager.m]: I thought you were going to persist the collection as a whole. Why the change of heart?

File: Signal/src/contact/OWSContactsManager.m [hunk 3]
```diff
@@ -240,170 +232,76 @@ - (void)updateSignalAccounts
...
-    });
+    return [OWSProfileManager sharedManager];
 }
 
 - (NSString *_Nullable)cachedDisplayNameForRecipientId:(NSString *)recipientId
 {
     OWSAssert(recipientId.length > 0);
 
-    return self.cachedAccountNameMap[recipientId];
+    SignalAccount *signalAccount = [self signalAccountForRecipientId:recipientId];
```
Comment (charlesmchen) [inline Signal/src/contact/OWSContactsManager.m]: `SignalAccount *_Nullable signalAccount` here and below...

File: SignalServiceKit/src/Contacts/SignalAccount.m [hunk 4]
```diff
@@ -46,6 +47,23 @@ - (nullable SignalRecipient *)signalRecipientWithTransaction:(YapDatabaseReadTra
...
+{
+    return _recipientId;
+}
+
+- (NSString *)displayName
+{
+    NSString *baseName = (self.contact.fullName.length > 0 ? self.contact.fullName : self.recipientId);
+
+    OWSAssert(self.hasMultipleAccountContact == (self.multipleAccountLabelText != nil));
+    NSString *displayName = (self.multipleAccountLabelText
```
Comment (charlesmchen) [inline SignalServiceKit/src/Contacts/SignalAccount.m]: This isn't redundant with logic in the `OWSContactsManager` is it?

File: SignalServiceKit/src/Contacts/SignalAccount.m [hunk 5]
```diff
@@ -46,6 +47,23 @@ - (nullable SignalRecipient *)signalRecipientWithTransaction:(YapDatabaseReadTra
     return [SignalRecipient recipientWithTextSecureIdentifier:self.recipientId withTransaction:transaction];
 }
 
+- (nullable NSString *)uniqueId
+{
+    return _recipientId;
```
Comment (charlesmchen) [inline SignalServiceKit/src/Contacts/SignalAccount.m]: I think we might want to check for empty recipientIds wherever we create new instances of `SignalRecipient`.

File: Signal/src/contact/OWSContactsManager.m (hunk 2 above)
Comment (michaelkirk) [reply Signal/src/contact/OWSContactsManager.m]: I started that way, but I realized there are still API's to look up the SignalRecipient by ID. If the cache wasn't been loaded when that happens we can produce the correct result by fetching them from disk directly.

File: SignalServiceKit/src/Contacts/SignalAccount.m (hunk 4 above)
Comment (michaelkirk) [reply SignalServiceKit/src/Contacts/SignalAccount.m]: This logic was moved from OWSContactsManager, but it lives only here now.

File: SignalServiceKit/src/Contacts/SignalAccount.m (hunk 5 above)
Comment (michaelkirk) [reply SignalServiceKit/src/Contacts/SignalAccount.m]: done

File: Signal/src/contact/OWSContactsManager.m (hunk 3 above)
Comment (michaelkirk) [reply Signal/src/contact/OWSContactsManager.m]: done
